src/access/accessFor.ts

```diff
@@ -0,0 +1,84 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { UrlString, WebId } from "../interfaces";
+import { internal_defaultFetchOptions } from "../resource/resource";
+import {
+  Access,
+  getAgentAccess,
+  getGroupAccess,
+  getPublicAccess,
+} from "./universal";
+
+export type Actor = "agent" | "group" | "public";
+
+/**
+ * Get an overview of what access is defined for a given actor (Agent, Group or everyone).
+ *
+ * This function works with Solid Pods that implement either the Web Access
+ * Control spec or the Access Control Policies proposal, with some caveats:
+ *
+ * - If access to the given Resource has been set using anything other than the
+ *   functions in this module, it is possible that it has been set in a way that
+ *   prevents this function from reliably reading access, in which case it will
+ *   resolve to `null`.
+ * - It will only return access specified explicitly for the given Agent. If
+ *   additional restrictions are set up to apply to the given Agent in a
+ *   particular situation, those will not be reflected in the return value of
+ *   this function.
+ * - It will only return access specified explicitly for the given Resource.
+ *   In other words, if the Resource is a Container, the returned Access may not
+ *   apply to contained Resources.
+ * - If the current user does not have permission to view access for the given
+ *   Resource, this function will resolve to `null`.
+ *
+ * @param resourceUrl URL of the Resource you want to read the access for.
+ * @param actorType type of actor whose access is being read.
+ * @param actor Identifier of the actor whose access being read for an Agent or a Group, undefined for everyone.
+ * @returns What access the given actor has.
+ */
+export async function getAccessFor(
+  resourceUrl: UrlString,
+  actorType: Actor,
+  actor?: WebId | UrlString,
+  options = internal_defaultFetchOptions
+): Promise<Access | null> {
```

Comment:
Might be nice to have some overloads here to narrow down which combinations of arguments are required:

```suggestion
export async function getAccessFor(
  resourceUrl: UrlString,
  actorType: "agent" | "group",
  actor: WebId | UrlString,
  options = internal_defaultFetchOptions
): Promise<Access | null>;
export async function getAccessFor(
  resourceUrl: UrlString,
  actorType: "public",
  options = internal_defaultFetchOptions
): Promise<Access | null>;
export async function getAccessFor(
  resourceUrl: UrlString,
  actorType: Actor,
  actor?: WebId | UrlString,
  options = internal_defaultFetchOptions
): Promise<Access | null>;
export async function getAccessFor(
  resourceUrl: UrlString,
  actorType: Actor,
  actor?: WebId | UrlString,
  options = internal_defaultFetchOptions
): Promise<Access | null> {
```

Actually... That brings to light that it doesn't work that way - right now, if you call it with `public`, you have to explicitly pass in `undefined` for `actor`. Maybe we could instead use `internal_defaultFetchOptions` as the default for `actor` as well, and then determine which to read depending on what argument is passed?

Reply:
I wasn't sure how to go with this: either the `actor` argument is part of the optional `options`, but it feels a bit different from the other functions in the API, or it is an argument on its own, but it must be assigned to `undefined` for `public` access. I went with the latter, but without a strong opinion, so I'll change this to the former since it looks like you prefer it from an API design point of view.

Reply:
I'm not sure if I prefer the former - I think `options` generally should catch optional arguments (hence the name :) ). However, if `actorType` is `public`, then what we call `actor` would ideally become the `options` object, if that makes sense?

Reply:
Fixed in b1a682995618590230c48aeb72a0e9ab9a71152f. The overload really makes it nicer, I had to add type assertions to pass in wrong parameters. With this, it actually makes more sense to have `actor` be part of `options`.

Also, I learned that default identifiers can't be used in overloads, only in the actual implementation (since it's technically not a type annotation)